### PR TITLE
Cooperative settlement

### PR DESCRIPTION
**Pull request by palango**
### What this PR does

Implement cooperative settle (https://github.com/raiden-network/raiden-contracts/issues/1383)

#### Tasks
- [x] Add check (+ test) for withdraw of to small amount (settle withdraw < deposit - withdraws)
- [x] Fix event data, adapt tests (`test_cooperative_settle_channel_event`)
- [x] Add gas estimates for cooperative settlement
- [x] Fix event checker to be able ti handle multiple events of same kind

### What's tricky about this PR (if any)

----

Any reviewer can check these:

* [ ] If the PR is fixing a bug or adding a feature, add an entry to CHANGELOG.md.
* [ ] If the PR changed a Solidity source, run `make compile_contracts` and add the resulting `raiden_contracts/data/contracts.json` in the PR.
* [ ] If the PR is changing documentation only, add `[skip ci]` in the commit message so Travis does not waste time.
    * [ ] But, if the PR changes comments in a Solidity source, do not add `[skip ci]` and let Travis check the hash of the source.
* [ ] In Python, use keyword arguments
* [ ] Squash unnecessary commits
* [ ] Comment commits
* [ ] Follow naming conventions
    * `solidityFunction`
    * `_solidity_argument`
    * `solidity_variable`
    * `python_variable`
    * `PYTHON_CONSTANT`
* [ ] Follow the Signature Convention in [CONTRIBUTING.md](./CONTRIBUTING.md)
* For each new contract
    * [ ] The deployment script deploys the new contract.
    * [ ] `etherscan_verify.py` runs on the new contract.
* Bookkeep
    * [ ] The gas cost of new functions are stored in `gas.json`.
* Solidity specific conventions
    * [ ] Document arguments of functions in natspec
    * [ ] Care reentrancy problems
    * if the PR adds or removes `require()` or `assert()`
        * [ ] add an entry in Changelog
        * [ ] open an issue in the client, light client, service repos so the change is reflected there
        * Just adding a message in `require()` doesn't require these steps.
* [ ] When you catch a require() failure in Solidity, look for a specific error message like `pytest.raises(TransactionFailed, match="error message"):`

And before "merge" all checkboxes have to be checked.  If you find redundant points, remove them.

### CODE REVIEW

File: raiden_contracts/data/source/raiden/TokenNetwork.sol [hunk 1]
```diff
@@ -479,7 +479,77 @@ contract TokenNetwork is Utils {
...
+        delete channel.participants[data2.participant];
+        delete channels[channel_identifier];
+
+        // Remove the pair's channel counter
+        pair_hash = getParticipantsHash(data1.participant, data2.participant);
+        delete participants_hash_to_channel_identifier[pair_hash];
 
+        // Emit channel lifecycle events
+        emit ChannelClosed(channel_identifier, data1.participant, 0, 0);
+        emit ChannelSettled(channel_identifier, data1.total_withdraw, 0, data2.total_withdraw, 0);
```
Comment (karlb) [inline raiden_contracts/data/source/raiden/TokenNetwork.sol:552]: Should we reuse the code from https://github.com/raiden-network/raiden-contracts/blob/15037b89cf322d1a1d309d097e06b7ed72c9e0b8/raiden_contracts/data/source/raiden/TokenNetwork.sol#L782-L809 here?

This question also shows that we are missing `storeUnlockData`. Is this a bug, or is there an explanation why we don't need it? We should add an explanation (if it is not needed) or a test to unlock after settlement (if it is needed).

File: raiden_contracts/tests/test_channel_cooperative_settle.py [hunk 2]
```diff
@@ -7,147 +7,202 @@
...
+            (B, balance_B, expiration, 0x0, signature_A2),
+        )
+    with pytest.raises(ValidationError):
+        token_network.functions.cooperativeSettle(
+            channel_identifier,
+            (A, balance_A, expiration, signature_A1, signature_B1),
+            (B, balance_B, expiration, signature_B2, 0x0),
+        )
+    # empty addresses
     with pytest.raises(TransactionFailed):
```
Comment (karlb) [inline raiden_contracts/tests/test_channel_cooperative_settle.py:91]: I like to check for the specific error with
```
with pytest.raises(TransactionFailed, match="error message"):
```
whenever possible, so that we know we are actually exercising the correct require. That does require error message, of course. But we should take the opportunity to add error messages (https://github.com/raiden-network/raiden-contracts/issues/81) before the next mainnet deployment now that #401 is done.

File: raiden_contracts/tests/test_channel_cooperative_settle.py [hunk 3]
```diff
@@ -312,13 +386,43 @@ def test_cooperative_settle_channel_state_withdraw(
...
+    (A, B, C) = get_accounts(3)
+    deposit_A = 20
+    deposit_B = 10
+    balance_A = 15 - 1
+    balance_B = 15
+
+    expiration = 100
+    channel_identifier = create_channel_and_deposit(A, B, deposit_A, deposit_B)
+
+    # Adding `withdraw_?` is neccessary, as `withdraw_amount` is monotonic increasing
```
Comment (karlb) [inline raiden_contracts/tests/test_channel_cooperative_settle.py:404]: I don't understand this comment.

File: raiden_contracts/tests/test_channel_cooperative_settle.py [hunk 4]
```diff
@@ -173,25 +230,28 @@ def test_cooperative_settle_channel_0(
...
-    create_cooperative_settle_signatures: Callable,
+    create_withdraw_signatures: Callable,
     cooperative_settle_state_tests: Callable,
 ) -> None:
     (A, B, C) = get_accounts(3)
     deposit_A = 0
     deposit_B = 0
     balance_A = 0
     balance_B = 0
+    expiration = 100
```
Comment (karlb) [inline raiden_contracts/tests/test_channel_cooperative_settle.py:246]: You could add `expiration=100` as a default to `create_withdraw_signatures` to make the tests a bit more concise.

File: raiden_contracts/tests/test_channel_cooperative_settle.py [hunk 5]
```diff
@@ -439,49 +562,73 @@ def test_cooperative_close_replay_reopened_channel(
...
     )
 
+    ev_handler.add(
+        txn_hash, ChannelEvent.WITHDRAW, check_withdraw(channel_identifier, B, balance_B)
+    )
+    # ev_handler.add(
+    #     txn_hash,
+    #     ChannelEvent.WITHDRAW,
+    #     check_withdraw(channel_identifier, A, balance_A)
+    # )
```
Comment (karlb) [inline raiden_contracts/tests/test_channel_cooperative_settle.py:628]: Why is this commented out? It looks like it should work.

File: raiden_contracts/tests/test_channel_cooperative_settle.py (hunk 4 above)
Comment (palango) [reply raiden_contracts/tests/test_channel_cooperative_settle.py:246]: Great idea!

Comment (palango) [reply raiden_contracts/tests/test_channel_cooperative_settle.py:246]: Mh, as it is also used in every contract call this doesn't help much. I put it into a constant instead.

File: raiden_contracts/tests/test_channel_cooperative_settle.py (hunk 5 above)
Comment (palango) [reply raiden_contracts/tests/test_channel_cooperative_settle.py:628]: The event checking code cannot handle checking two events with the same name in the same tx right now. I didn't want to implement that before a first review on the contract changes.

File: raiden_contracts/tests/test_channel_cooperative_settle.py (hunk 2 above)
Comment (palango) [reply raiden_contracts/tests/test_channel_cooperative_settle.py:93]: Done for all tests where transaction are expected to fail. I didn't add it to the `web3` formatting tests. Do you think it's useful there as well?

Comment (karlb) [reply raiden_contracts/tests/test_channel_cooperative_settle.py:93]: I don't think so. IMO, those have limited value, anyway.

File: raiden_contracts/data/source/raiden/TokenNetwork.sol (hunk 1 above)
Comment (palango) [reply raiden_contracts/data/source/raiden/TokenNetwork.sol]: I factored out the mentioned lines and added some comments about why this doesn't allow remaining locks in the docstring as well as the code.

File: raiden_contracts/data/source/raiden/TokenNetwork.sol [hunk 6]
```diff
@@ -479,7 +506,74 @@ contract TokenNetwork is Utils {
...
+                data2.total_withdraw,
+                data2.expiration_block,
+                data2.participant_signature,
+                data2.partner_signature
+            );
+        }
+        removeChannelData(channel, channel_identifier, data1.participant, data2.participant);
+
+        // Emit channel lifecycle events
+        emit ChannelClosed(channel_identifier, data1.participant, 0, 0);
```
Comment (karlb) [inline raiden_contracts/data/source/raiden/TokenNetwork.sol:575]: I'm not sure if we should emit this event, since the channel will never be in a closed state. I think I prefer to have it to make both settlement approaches more similar, but it does not match the concept fully (the nonce and balance_hash do not really apply) and removing it would also be slightly cheaper.

**Review by karlb — APPROVED** [review]
Really nice to finally have a coop settle after such a long time!

File: raiden_contracts/data/source/raiden/TokenNetwork.sol (hunk 6 above)
Comment (palango) [reply raiden_contracts/data/source/raiden/TokenNetwork.sol:575]: The intention was to keep the same channel lifecycle as before. In other words, I am not sure if the PFS/explorer/... rely on the close event in some sense.

But I will check on this again.

**Review by andrevmatos — APPROVED** [review]
I'm no solidity expert, but this looks good to me. It'd be good to have someone else more into solidity itself to review it though. I like the approach to leverage the double-withdraw signed structures, but we really need to optimise those gas usages

File: raiden_contracts/data/source/raiden/TokenNetwork.sol (hunk 6 above)
Comment (palango) [reply raiden_contracts/data/source/raiden/TokenNetwork.sol:571]: Seems like we need some adjustments in either case (with or without the `ChannelClosed` event):

- The PFS currently relies on the `ChannelClosed` event, and doesn't even handle the `ChannelSettled` event:
https://github.com/raiden-network/raiden-services/blob/37c6f4816e2a6a084bd4e7e9c0eb58b501870b66/src/pathfinding_service/service.py#L228-L229
  - However, adding the same handling for `ChannelSettled` events shouldn't be difficult
- The MS triggers a `ActionMonitoringTriggeredEvent` (https://github.com/raiden-network/raiden-services/blob/37c6f4816e2a6a084bd4e7e9c0eb58b501870b66/src/monitoring_service/handlers.py#L124) when a channel is closed. The event handler (https://github.com/raiden-network/raiden-services/blob/37c6f4816e2a6a084bd4e7e9c0eb58b501870b66/src/monitoring_service/handlers.py#L435) for this event doesn't currently checked if the channel in question is settled already.


tldr: We need to adjust the services anyways, so we can decide freely on which approach makes more logical sense.

File: raiden_contracts/data/source/raiden/TokenNetwork.sol [hunk 7]
```diff
@@ -475,7 +502,74 @@ contract TokenNetwork is Utils {
...
+        require(data1.total_withdraw <= data1.total_withdraw + data2.total_withdraw);
+
+        if (data1.total_withdraw > 0) {
+            this.setTotalWithdraw(
+                channel_identifier,
+                data1.participant,
+                data1.total_withdraw,
+                data1.expiration_block,
+                data1.participant_signature,
+                data1.partner_signature
```
Comment (andrevmatos) [inline raiden_contracts/data/source/raiden/TokenNetwork.sol:555]: Nitpick question: can't we call `setTotalWithdraw2` directly with `data1` here? And data2 below? Wouldn't this be cheaper in gas?

Comment (palango) [reply raiden_contracts/data/source/raiden/TokenNetwork.sol]: Well spotted!

File: raiden_contracts/data/source/raiden/TokenNetwork.sol (hunk 6 above)
Comment (palango) [reply raiden_contracts/data/source/raiden/TokenNetwork.sol:563]: We agreed to not emit the `ChannelClosed` event.
